### DISCUSSION TITLE
Sampling PGO support

### DISCUSSION THREAD

**Discussion by zamazan4ik** [oracle/graal#7648] [Ideas] [answered]
Hi!

According to the GraalVM [documentation](https://www.graalvm.org/22.0/reference-manual/native-image/PGO/), GraalVM supports PGO only with Instrumentation. However, other AoT compilers like Clang, GCC, Rustc support an additional PGO kind - Sampling PGO (also called [AutoFDO](https://github.com/google/autofdo)).

Is there a way to use anyhow sampling PGO with GraalVM? If not, are there plans for implementing sampling PGO in GraalVM?

Having sampling PGO support is important for many PGO-in-production scenarios, especially if we are talking about using PGO at large scales. More about this topic you can find in Google's paper: https://research.google/pubs/pub45290/

Links:
* Sampling PGO in Clang (the same applies to Rustc since Clang and Rustc are both LLVM-based): [docs](https://clang.llvm.org/docs/UsersManual.html#using-sampling-profilers)
* Sampling PGO in GCC: [GCC Wiki](https://gcc.gnu.org/wiki/AutoFDO/Tutorial)

**Comment by axel22** [comment, marked as answer]
Yes, the PGO in GraalVM EE does both instrumentation and sampling in the last two releases.
The call-stack samples are included in the `.iprof` files next to the instrumentation profiles.
On nightlies (didn't check if it made it to the last release), there is an option `-H:+BuildReport -H:+BuildReportSamplerFlamegraph` that will render these samples inside the build report file (so that they can be inspected visually in a flame graph).
The samples are used to improve some of the EE optimizations.

We will improve our official documentation to explain this better.

**Reply by zamazan4ik** [reply]
Is it possible in the current sampling PGO implementation to tweak the sampling frequency (like it could be done with `perf` in the AutoFDO case)?

It's important for us since we can tune the "Sampling overhead <-> Profile precision" tradeoff.

Another question - according to your tests, are there differences between instrumentation and sampling PGO from the optimization opportunity perspective? Because Google reports AutoFDO (sampling PGO) optimization is a bit worse compared to the instrumentation PGO. I am wondering how true it is for the GraalVM PGO implementation.

**Reply by zamazan4ik** [reply]
@axel22 friendly ping :)
